src/service/xhr-impl.js

```diff
@@ -125,16 +126,146 @@ export class Xhr {
     dev().assert(
         creds === undefined || creds == 'include' || creds == 'omit',
         'Only credentials=include|omit support: %s', creds);
+
+    if (this.ampdocSingle_) {
+      const viewer = Services.viewerForDoc(this.ampdocSingle_);
+      if ((viewer.isTrustedViewer() || getMode().development) &&
+          viewer.hasCapability('xhrInterceptor')) {
+        const messagePayload = {
+          originalRequest: this.toStructuredCloneable_(input, init),
+        };
+
+        return viewer.sendMessageAwaitResponse('xhr', messagePayload)
+            .then(response => this.fromStructuredCloneable_(response));
```

Comment:
We should catch viewer errors (promise rejections).

Reply:
How should we handle viewer errors? I was thinking to propagate viewer errors to the callers as if it was an XHR error. Do we maybe want to transform any non-XHR viewer-specific errors to an appropriate abstracted error?